File: src/algorithm/leetcode/Problem538.java
package algorithm.leetcode;

public class Problem538 {
    public TreeNode convertBST(TreeNode root) {
        traverse(root, 0);
        return root;
    }

    public static int traverse(TreeNode root, int greaterValue) {
        if (root == null)
            return 0;
        int right = traverse(root.right, greaterValue);
        int left = traverse(root.left, greaterValue + right + root.val);
        int val = right + left + root.val;
        root.val += greaterValue + right;
        return val;
    }
}

File: src/algorithm_cpp/problem538.cpp
class Solution {
public:
    TreeNode* search(TreeNode* curnode, int& curval) {
	if (curnode == nullptr) return nullptr;
	TreeNode* right = search(curnode->right, curval);
	curval += curnode->val;
	TreeNode* cur = new TreeNode(curval);
	TreeNode* left = search(curnode->left, curval);
	cur->left = left;
	cur->right = right;
	return cur;
    }

    TreeNode* convertBST(TreeNode* root) {
	int curval = 0;
	return search(root, curval);
    }
};
